_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  application code. However, there is more weight and importance in the review process here as merging this code will
+  update the live infrastructure.
+
+Run plan automatically::
+  As mentioned above, there is very little automated testing you can do with the live config. As such, the only form of
+  automation you can add to the review process is to do a dry run of the infrastructure and make it available so that
+  the reviewer can take a look.
+
+Merge and deploy::
+  Once the code has been reviewed and the plan makes sense, it is time to merge and deploy the change. Given the _Golden
+  Rule of Terraform_, where the master should be a 1:1 representation of what is actually deployed, the live
+  configuration typically does not have a release process. That is, typically you do not cut a release and artifact the
+  code. Instead, you design the CI/CD workflow so that on merge to master the code is immediately scheduled for
+  deployment. However, just like with any other code, you can introduce subtle integration bugs in the merge process.
+  Just looking at the plan from the feature branch PR is not sufficient to automatically run deploy the configuration
+  after merging the code since there is no guarantee that the exact same plan will be produced after the code has merged
+  into trunk. This is because other changes not available on your feature branch may have been made in the meantime,
+  causing both the infrastructure and code to change. Therefore it is always important to rerun the plan before
+  deploying the infrastructure, and having an approval process baked into the CD pipeline itself. That is, your
+  automated pipeline should:
+
+    1. Do a dry run (`terraform plan`) from the updated trunk.
+    1. Notify that a deployment is scheduled and a plan is available for review.
+    1. Wait for manual approval.
+    1. Deploy the code only after it has been reviewed for correctness.
+
+In summary, here are the key differences with live infrastructure configurations and the other two kinds of code we have
+discussed:
+
+- There is almost no form of automated testing you can implement in the predeployment stage. The only thing you can do
+  is perform a dry run and review the plan.
+
+- Similarly, there is no alternative environment where you can test the code manually during development; not even a
+  sandbox environment.
+
+- There are no release artifacts or tag with the live code. Everything is deployed immediately after reaching trunk.
+
+Now that we have an idea of what CI/CD pipelines look like, let's take a step back and define a threat model for CI/CD.
+This threat model will help us ensure that we implement the necessary security controls in these CI/CD pipelines so that
+we cover the common types of attack vectors for this type of workflow.
+
+
+[[threat_model_of_cicd]]
+=== Threat model of CI/CD
+
+To implement a CI/CD pipeline for infrastructure code, it is required that the ultimate entity or system running the
```

Comment:
The article is on CI / CD in general, but this first sentence focuses on infra code. Might want to compare app and infra code here side by side again. Also shows the far more limited permissions necessary to deploy an app.

Reply:
Updated bc14490